src/librustc/traits/object_safety.rs

```diff
@@ -323,9 +316,108 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         }
 
+        let receiver_ty = self.liberate_late_bound_regions(
+            method.def_id,
+            &sig.map_bound(|sig| sig.inputs()[0]),
+        );
+
+        // until we get by-value DST, `self: Self` can't be coerced
+        // but we allow this as a special case.
+        // maybe instead we should also check for
+        // for (U) { if (Self: Unsize<U>) { Receiver: Unsize<Receiver<Self=Self>>}}
+        if receiver_ty != self.mk_self_type() {
+            if !self.receiver_is_coercible(method, receiver_ty) {
+                return Some(MethodViolationCode::UncoercibleReceiver);
+            }
+        }
+
         None
     }
 
+    // checks the type of the self argument, and makes sure it implements
+    // the CoerceUnsized requirement:
+    // forall (U) {
+    //     if (Self: Unsize<U>) {
+    //         Receiver: CoerceUnsized<Receiver<Self=U>>
+    //     }
+    // }
+    #[allow(dead_code)]
+    fn receiver_is_coercible(
+        self,
+        method: &ty::AssociatedItem,
+        receiver_ty: Ty<'tcx>,
+    ) -> bool
+    {
+        debug!("receiver_is_coercible: method = {:?}, receiver_ty = {:?}", method, receiver_ty);
+
+        let traits = (self.lang_items().unsize_trait(),
+                      self.lang_items().coerce_unsized_trait());
+        let (unsize_did, coerce_unsized_did) = if let (Some(u), Some(cu)) = traits {
+            (u, cu)
+        } else {
+            debug!("receiver_is_coercible: Missing Unsize or CoerceUnsized traits");
+            return false;
+        };
+
+        // use a bogus type parameter to mimick a forall(U) query
+        // using u32::MAX for now. This is BAD and will probably break when
+        // this method is called recursively, or at least if someone does a hacky thing
+        // like this elsewhere in the compiler
+        let target_self_ty: Ty<'tcx> = self.mk_ty_param(
+            ::std::u32::MAX,
+            Name::intern("mikeyhewROCKS").as_interned_str(),
+        );
+
+        // create a modified param env, with
+        // `Self: Unsize<U>` added to the caller bounds
+        let param_env = {
+            let mut param_env = self.param_env(method.def_id);
+
+            let predicate = ty::TraitRef {
+                def_id: unsize_did,
+                substs: self.mk_substs_trait(self.mk_self_type(), &[target_self_ty.into()]),
+            }.to_predicate();
+
+            let caller_bounds: Vec<Predicate<'tcx>> = param_env.caller_bounds.iter().cloned()
+                .chain(iter::once(predicate))
+                .collect();
+
+            param_env.caller_bounds = self.intern_predicates(&caller_bounds);
+
+            param_env
+        };
+
+        let receiver_substs = Substs::for_item(self, method.def_id, |param, _| {
+            if param.index == 0 {
+                target_self_ty.into()
+            } else {
+                self.mk_param_from_def(param)
+            }
+        });
+
+        // the type `Receiver<Self=U>` in the query
```

Comment:
Nit: I think I would denote this as `[Self => U] Receiver` -- i.e., we are applying a substitution, right?

Reply:
Or maybe `Receiver<Self => U>`...

Reply:
You're right. I was thinking that `Self=U` made sense because `Self` is a type parameter, but it's a type parameter for the trait, not the receiver type, so it really is just a substitution.

The reason I shied away from using substitution notation is I always forget which way it goes 😅 does `[X/Y]` mean `X` replaced with `Y` or `Y` replaced with `X`?

Reply:
This is why I tend to use an arrow like `=>` =)

(For reference, though, `[Y/X]` is "substitutes Y *for* X" -- i.e., `[X => Y]`)